src/Symfony/Component/Utf8/Bytes.php

```diff
@@ -0,0 +1,505 @@
+        }
+
+        if (false === $part = strstr($this->string, $needle, $beforeNeedle)) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = $part;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function substringOfIgnoreCase(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        $current = setlocale(LC_CTYPE, '0');
+        setlocale(LC_CTYPE, 'C');
+        $part = stristr($this->string, $needle, $beforeNeedle);
+        setlocale(LC_CTYPE, $current);
+
+        if (false === $part) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = $part;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastSubstringOf(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (!$beforeNeedle) {
+
+            if (false === $part = strrchr($this->string, $needle)) {
+                return;
+            }
+
+            $result = clone $this;
+            $result->string = $part;
+
+            return $result;
+        }
+
+        if (false === $offset = strrpos($this->string, $needle)) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = substr($this->string, 0, $offset);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function lastSubstringOfIgnoreCase(string $needle, bool $beforeNeedle = false)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (false === $offset = strripos($this->string, $needle)) {
+            return;
+        }
+
+        $result = clone $this;
+
+        $current = setlocale(LC_CTYPE, '0');
+        setlocale(LC_CTYPE, 'C');
+        $result->string = $beforeNeedle ? substr($this->string, 0, $offset) : substr($this->string, $offset);
+        setlocale(LC_CTYPE, $current);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reverse()
+    {
+        $result = clone $this;
+        $result->string = strrev($this->string);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function width(bool $ignoreAnsiDecoration = true): int
+    {
+        $s = $this->string;
+
+        if (false !== strpos($s, "\r")) {
+            $s = str_replace("\r\n", "\n", $s);
+            $s = strtr($s, "\r", "\n");
+        }
+
+        $width = 0;
+        foreach (explode("\n", $s) as $s) {
+            if ($ignoreAnsiDecoration) {
+                $s = preg_replace('/\x1B\[[\d;]*m/', '', $s);
+            }
+
+            if ($width < $c = strlen($s)) {
+                $width = $c;
+            }
+        }
+
+        return $width;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replace($from, $to, int &$count = null)
```

Comment:
A tuple would mean:

```php
public function replace($from, $to): array {
    // validation ...
    $str = str_replace($from, $to, $str, $count);

    return [$str, $count];
}

$str = $str->replace('from', 'to')[0];
```

This kind of unpacking might seem strange to PHP developers, but is very common in other languages. Just think of Go where people are constantly unpacking. Of course those languages provide better means to deal with such constructs.

A result class should be obvious, it could be as simple as an extended `Bytes` instance that has a `getReplaceCount` (or similar). Hence, no additional unpacking is required to continue working with it.

Regarding in-out parameters: https://softwareengineering.stackexchange.com/questions/245767